source/extensions/transport_sockets/tls/context_impl.cc

```diff
@@ -1314,17 +1384,63 @@ bool ServerContextImpl::isClientEcdsaCapable(const SSL_CLIENT_HELLO* ssl_client_
   return false;
 }
 
+OcspStapleAction ServerContextImpl::passesOcspPolicy(const ContextImpl::TlsContext& ctx) {
+  auto& response = ctx.ocsp_response_;
+  if (!Runtime::runtimeFeatureEnabled(
+          "envoy.reloadable_features.validate_ocsp_expiration_on_connection")) {
+    return response ? OcspStapleAction::Staple : OcspStapleAction::NoStaple;
+  }
+
+  if (ctx.is_must_staple_ && response && response->isExpired()) {
+    return OcspStapleAction::Fail;
+  }
+
+  if (response && !response->isExpired()) {
+    return OcspStapleAction::Staple;
+  }
+
+  // `response` is either not present or expired, and the action depends on the policy.
+  switch (ocsp_staple_policy_) {
+  case Ssl::ServerContextConfig::OcspStaplePolicy::SkipStaplingIfExpired:
+    return OcspStapleAction::NoStaple;
+  case Ssl::ServerContextConfig::OcspStaplePolicy::StaplingRequired:
+    return OcspStapleAction::Fail;
+  case Ssl::ServerContextConfig::OcspStaplePolicy::RejectConnectionOnExpired:
+    return response ? OcspStapleAction::Fail : OcspStapleAction::NoStaple;
+  default:
+    NOT_REACHED_GCOVR_EXCL_LINE;
+  }
+}
+
 enum ssl_select_cert_result_t
 ServerContextImpl::selectTlsContext(const SSL_CLIENT_HELLO* ssl_client_hello) {
   const bool client_ecdsa_capable = isClientEcdsaCapable(ssl_client_hello);
   // Fallback on first certificate.
   const TlsContext* selected_ctx = &tls_contexts_[0];
   for (const auto& ctx : tls_contexts_) {
-    if (client_ecdsa_capable == ctx.is_ecdsa_) {
+    if (client_ecdsa_capable == ctx.is_ecdsa_ && passesOcspPolicy(ctx) != OcspStapleAction::Fail) {
       selected_ctx = &ctx;
       break;
     }
   }
+
+  switch (passesOcspPolicy(*selected_ctx)) {
+  case OcspStapleAction::Staple: {
+    RELEASE_ASSERT(selected_ctx->ocsp_response_,
+                   "OCSP response must be present under OcspStapleAction::Staple");
+    auto& resp_bytes = selected_ctx->ocsp_response_->rawBytes();
+    RELEASE_ASSERT(
+        SSL_set_ocsp_response(ssl_client_hello->ssl, resp_bytes.data(), resp_bytes.size()), "");
```

Comment:
Is it safe to do this even if the client didn't request a staple? Should this be done in the other callback you added above?

Should OCSP policy affect cert selection if the client didn't request a staple? (That's hard to implement though, and probably not worth changing; I'm just thinking out loud).

Reply:
It's safe to do because boringSSL will only transmit the stapled response if the client requests it (but it's not possible to know at this point whether the client will request it or not).

The OCSP policy only prevents selection of certs where we would fail the connection on the OCSP policy. In that case we try later certs (though in practice I think it's only exactly 1 or 2 certs).

Reply:
Can you add a comment, saying that this will only be attached if the client asks for it?

Reply:
I went ahead and reworked this to not require the client to request OCSP stapling. The reason being that providing a cert with the must-staple extension upgrades the policy to MUST_STAPLE, but it's not guaranteed that all clients will support OCSP. Seems like they probably shouldn't be denied access because of it.